addons/webhooks/clusterbootstrap_webhook.go

```diff
@@ -0,0 +1,442 @@
+	ticker := time.NewTicker(constants.DiscoveryCacheInvalidateInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			wh.cachedDiscoveryClient.Invalidate()
+			wh.providerGVR = make(map[schema.GroupKind]*schema.GroupVersionResource)
+		}
+	}
+}
+
+// validateMandatoryCorePackageUpdate combines new package spec validation and package upgrade validation together
+func (wh *ClusterBootstrap) validateMandatoryCorePackageUpdate(ctx context.Context, old, new *runv1alpha3.ClusterBootstrapPackage, namespace string, fldPath *field.Path) *field.Error {
+	if err := wh.validateClusterBootstrapPackage(ctx, new, namespace, fldPath); err != nil {
+		return err
+	}
+	if err := wh.validateClusterBootstrapPackageUpdate(ctx, old, new, fldPath); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (wh *ClusterBootstrap) validateOptionalCorePackageUpdate(ctx context.Context, old, new *runv1alpha3.ClusterBootstrapPackage, namespace string, fldPath *field.Path) *field.Error {
+	if new != nil {
+		if err := wh.validateClusterBootstrapPackage(ctx, new, namespace, fldPath); err != nil {
+			return err
+		}
+		if old != nil {
+			if err := wh.validateClusterBootstrapPackageUpdate(ctx, old, new, fldPath); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (wh *ClusterBootstrap) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	// Covert objs to ClusterBootstrap
+	newClusterBootstrap, ok := newObj.(*runv1alpha3.ClusterBootstrap)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterBootstrap but got a %T", newObj))
+	}
+	oldClusterBootstrap, ok := oldObj.(*runv1alpha3.ClusterBootstrap)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterBootstrap but got a %T", oldObj))
+	}
+	clusterbootstraplog.Info("validate update", "name", newClusterBootstrap.Name)
+
+	var allErrs field.ErrorList
+	namespace := newClusterBootstrap.Namespace
+	if err := wh.validateMandatoryCorePackageUpdate(ctx, oldClusterBootstrap.Spec.CNI, newClusterBootstrap.Spec.CNI, namespace, getFieldPath("cni")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := wh.validateMandatoryCorePackageUpdate(ctx, oldClusterBootstrap.Spec.Kapp, newClusterBootstrap.Spec.Kapp, namespace, getFieldPath("kapp")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	// CSI and CPI can be nil
+	if err := wh.validateOptionalCorePackageUpdate(ctx, oldClusterBootstrap.Spec.CSI, newClusterBootstrap.Spec.CSI, namespace, getFieldPath("csi")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := wh.validateOptionalCorePackageUpdate(ctx, oldClusterBootstrap.Spec.CPI, newClusterBootstrap.Spec.CPI, namespace, getFieldPath("cpi")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	// validate new additional packages
+	addtionalPkgFldPath := getFieldPath("additionalPackages")
```

Comment:
Does this need to be changed to `getFieldPath("additionalPackages").Child(string(rune(idx)))` like above?

Reply:
Thanks for pointing out. Changed for all the additional packages.

Reply:
Is there a test to exercise this flow?

Reply:
There is a testcase for `// Additional package can't be removed`, all other logics in validating spec should be the same as core packages, as they are calling the same functions.

Reply:
Had an offline sync. fieldPath is used in error messages and not for the actual object.